aws/amazon-cloudwatch-agent
# Fix missing Windows events by changing getRecords().

# Description of the issue
Closes https://github.com/aws/amazon-cloudwatch-agent/issues/199
CloudWatch Agent reads from the Windows event logs in batches.
Currently, if the Agent has issues reading or rendering _any_ event in the batch, it will drop all the events in the batch.
In this specific case there was an event related to the install of Google Chrome for which `EvtOpenPublisherMetadata()` failed.

# Description of changes
Change `getRecords()` to `continue` when an error is encountered, instead of `return`.
Added unit tests.
Fixed another bug in `windowsEventLog.Open()` so it returns an error when `EvtSubscribe()` fails.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Manual test:
1. Start an EC2 instance running Windows Server 2019 with the latest CloudWatch Agent installed and configured to monitor the Application Event log for ERROR, INFO, VERBOSE.
2. Run a PowerShell script which repeatedly writes an ERROR level event to the Application event log with a specific message and an increasing counter.
3. Install and uninstall Google Chrome.
4. Expect CloudWatch console to show all events from the PowerShell script, but actual result shows missing events.

Added unit test:
```
PS C:\Users\Administrator\go\src\amazon-cloudwatch-agent> go test  .\plugins\inputs\windows_event_log\wineventlog\...
ok      github.com/aws/amazon-cloudwatch-agent/plugins/inputs/windows_event_log/wineventlog     6.594s
```
